celery_pubsub/pubsub.py

```diff
@@ -1,61 +1,37 @@
+import re
 import celery
 
-__all__ = ['publish', 'publish_now', 'subscribe', 'unsubscribe', ]
-
 
 class PubSubManager(object):
+
     def __init__(self):
-        super(PubSubManager, self).__init__()
-        self.subscribed = []
         self.jobs = {}
 
     def publish(self, topic, *args, **kwargs):
-        result = self.get_jobs(topic).delay(*args, **kwargs)
-        return result
+        group = []
+        for (regex, task) in self.jobs.values():
+            if regex.match(topic):
+                group.append(task)
+        out = celery.group(group)
+        return out
 
     def publish_now(self, topic, *args, **kwargs):
-        result = self.get_jobs(topic).apply(args=args, kwargs=kwargs)
+        # REVIEW: useless shortcut?
+        result = self.publish(topic, *args, **kwargs).apply(args=args, kwargs=kwargs)
         return result
 
     def subscribe(self, topic, task):
-        regex = self._topic_to_re(topic)
-        if (topic, regex, task) not in self.subscribed:
-            self.subscribed.append((topic, regex, task))
-            self.jobs = {}
-
-    def unsubscribe(self, topic, task):
-        found = -1
-        for idx, triplet in enumerate(self.subscribed):
-            if (triplet[0], triplet[2]) == (topic, task):
-                found = idx
-                break
-        if not found == -1:
-            del self.subscribed[found]
-            self.jobs = {}
-
-    def get_jobs(self, topic):
-        if topic not in self.jobs:
-            self._gen_jobs(topic)
-        return self.jobs[topic]
-
-    def _gen_jobs(self, topic):
-        jobs = []
-        for job in self.subscribed:
-            if job[1].match(topic):
-                jobs.append(job[2].s())
-        self.jobs[topic] = celery.group(jobs)
```

Comment:
`_gen_jobs` does not generate a job, it creates a celery group with the possibly hierarchically organized list of procedures.

Reply:
Well, it generates `self.jobs` for its key that is currently being requested.

Reply:
And no hierarchy. Only the ones that have been found to match the topic :-).

Reply:
`topic` is organised using a dotted expression that describe a hierarchy.

Reply:
Right, semantics. I meant that the structure of `self.jobs[xxx]` is no longer a hierarchy but flat. I think this can be resolved :-)